### ISSUE TITLE
Excessive loss of precision

### ISSUE THREAD

**Issue by proteneer** [google/jax#641]
I'm seeing an excessive loss of precision far beyond what is typically acceptable for 64-bit calculations:

``` python
from jax.config import config; config.update("jax_enable_x64", True)

import jax
import jax.numpy as np
import numpy as anp
from jax.test_util import check_grads
import unittest

class PeriodicTorsion():

    def __init__(self,
        torsion_idxs,
        param_idxs):
        """
        This implements a periodic torsional potential expanded out into three terms:

        V(a) = k0*(1+cos(1 * a - t0)) + k1*(1+cos(2 * a - t1)) + k2*(1+cos(3 * a - t2))

        Parameters:
        -----------
        torsion_idxs: [num_torsions, 4] np.array
            each element (a, b, c, d) is a torsion of four atoms, defined as
            as the angle of the plane defined by the three bond vectors a-b, b-c, c-d. 

        param_idxs: [num_torsions, 6] np.array
            each element (k, phase, periodicity) maps into params for angle constants and ideal angles

        """
        self.torsion_idxs = torsion_idxs
        self.param_idxs = param_idxs

    @staticmethod
    def get_signed_angle(ci, cj, ck, cl):
        """
        The torsion angle between two planes should be periodic but not
        necessarily symmetric. We use an identical but numerically stable arctan2
        implementation as opposed to the OpenMM energy function to avoid a
        singularity when the angle is zero. This is different from the CHARMM convention
        of symmetric torsions.
        """

        # Taken from the wikipedia arctan2 implementation:
        # https://en.wikipedia.org/wiki/Dihedral_angle

        rij = ci - cj
        rkj = ck - cj
        rkl = ck - cl

        n1 = np.cross(rij, rkj)
        n2 = np.cross(rkj, rkl)

        lhs = np.linalg.norm(n1, axis=-1)
        rhs = np.linalg.norm(n2, axis=-1)
        bot = lhs * rhs

        y = np.sum(np.multiply(np.cross(n1, n2), rkj/np.linalg.norm(rkj, axis=-1, keepdims=True)), axis=-1)
        x = np.sum(np.multiply(n1, n2), -1)

        return np.arctan2(y, x)

    def angles(self, conf):
        ci = conf[self.torsion_idxs[:, 0]]
        cj = conf[self.torsion_idxs[:, 1]]
        ck = conf[self.torsion_idxs[:, 2]]
        cl = conf[self.torsion_idxs[:, 3]]

        angle = self.get_signed_angle(ci, cj, ck, cl)
        return angle

    def energy(self, conf, params):
        """
        Compute the torsional energy.
        """
        ci = conf[self.torsion_idxs[:, 0]]
        cj = conf[self.torsion_idxs[:, 1]]
        ck = conf[self.torsion_idxs[:, 2]]
        cl = conf[self.torsion_idxs[:, 3]]

        ks = params[self.param_idxs[:, 0]]
        phase = params[self.param_idxs[:, 1]]
        period = params[self.param_idxs[:, 2]]
        angle = self.get_signed_angle(ci, cj, ck, cl)
        nrg = ks*(1+np.cos(period * angle - phase))
        return np.sum(nrg, axis=-1)


class TestPeriodicTorsion(unittest.TestCase):

    def setUp(self):
        self.conformers = anp.array([
            [[-0.6000563454193615, 0.376172954382274 ,-0.2487295756125901],
             [ 0.561317027011325 , 0.2066950040043141, 0.3670430960815993],
             [-1.187055522272264 ,-0.3415864358441354, 0.0871382207830652],
             [ 0.9399773448903637,-0.6888774474110431, 0.2104211949995816]],
            [[-0.6000563454193615, 0.376172954382274 ,-0.2487295756125901],
             [ 0.5613170270113252, 0.2066950040043142, 0.3670430960815993],
             [-1.187055522272264 ,-0.3415864358441354, 0.0871382207830652],
             [ 1.283345455745044 ,-0.0356257425880843,-0.2573923896494185]],
            [[-0.6000563454193615, 0.376172954382274 ,-0.2487295756125901],
             [ 0.561317027011325 , 0.2066950040043142, 0.3670430960815992],
             [-1.187055522272264 ,-0.3415864358441354, 0.0871382207830652],
             [ 1.263820400176392 , 0.7964992122869241, 0.0084568741589791]],
            [[-0.6000563454193615, 0.376172954382274 ,-0.2487295756125901],
             [ 0.5613170270113252, 0.2066950040043142, 0.3670430960815992],
             [-1.187055522272264 ,-0.3415864358441354, 0.0871382207830652],
             [ 0.8993534242298198, 1.042445571242743 , 0.7635483993060286]],
            [[-0.6000563454193615, 0.376172954382274 ,-0.2487295756125901],
             [ 0.5613170270113255, 0.2066950040043142, 0.3670430960815993],
             [-1.187055522272264 ,-0.3415864358441354, 0.0871382207830652],
             [ 0.5250337847650304, 0.476091386095139 , 1.3136545198545133]],
            [[-0.6000563454193615, 0.376172954382274 ,-0.2487295756125901],
             [ 0.5613170270113255, 0.2066950040043141, 0.3670430960815993],
             [-1.187055522272264 ,-0.3415864358441354, 0.0871382207830652],
             [ 0.485009232042489 ,-0.3818599172073237, 1.1530102055165103]],
            ], dtype=anp.float64)

        self.nan_conformers = anp.array([
            [[-0.6000563454193615, 0.376172954382274 ,-0.2487295756125901],
             [ 0.5613170270113252, 0.2066950040043142, 0.3670430960815993],
             [-1.187055522272264 ,-0.3415864358441354, 0.0871382207830652],
             [ 1.2278668040866427, 0.8805184219394547, 0.099391329616366 ]],
            [[-0.6000563454193615, 0.376172954382274 ,-0.2487295756125901],
             [ 0.561317027011325 , 0.206695004004314 , 0.3670430960815994],
             [-1.187055522272264 ,-0.3415864358441354, 0.0871382207830652],
             [ 0.5494071252089705,-0.5626592973923106, 0.9817919758125693]],
            ], dtype=anp.float64)


    def test_gpu_torsions(self):
        """
        Test agreement of torsions with OpenMM's implementation of torsion terms.
        """
        torsion_idxs = anp.array([
            [0, 1, 2, 3],
            # [0, 1, 2, 3],
            # [0, 1, 2, 3],
        ], dtype=anp.int32)

        params_np = anp.array([
            2.3, # k0
            5.4, # k1
            9.0, # k2
            0.0, # t0
            3.0, # t1
            5.8, # t2
            1.0, # n0
            2.0, # n1
            3.0  # n2
        ], dtype=anp.float64)

        param_idxs = anp.array([
            # [0, 3, 6], # works
            [1, 4, 7], # fails
            # [2, 5, 8] # fails
        ], dtype=anp.int32)

        ref_nrg = PeriodicTorsion(
            param_idxs=param_idxs,
            torsion_idxs=torsion_idxs
        )

        for conf_idx, conf in enumerate(self.conformers):
            nrg = ref_nrg.energy(conf, params_np)
            angles = ref_nrg.angles(conf)

            print(nrg, angles)

            # first order derivatives pass
            check_grads(ref_nrg.angles, (conf,), order=1) # passes
            check_grads(ref_nrg.energy, (conf, params_np), order=1) # passes


            # second order derivatives fail miserably, both of these should pass in 64bit
            check_grads(ref_nrg.angles, (conf,), order=2) # fails
            check_grads(ref_nrg.energy, (conf, params_np), order=2) # fails




if __name__ == "__main__":
    unittest.main()
```

The failures look like:

```
======================================================================
FAIL: test_gpu_torsions (__main__.TestPeriodicTorsion)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "timemachine/jax_functionals/loss_of_precision_repro.py", line 175, in test_gpu_torsions
    check_grads(ref_nrg.energy, (conf, params_np), order=2) # fails
  File "/Users/hessian/venv/lib/python3.6/site-packages/jax/test_util.py", line 153, in check_grads
    check_grads(f_vjp, args, order - 1, atol=atol, rtol=rtol, eps=eps)
  File "/Users/hessian/venv/lib/python3.6/site-packages/jax/test_util.py", line 160, in check_grads
    check_vjp(f, partial(api.vjp, f), args, atol, rtol, eps)
  File "/Users/hessian/venv/lib/python3.6/site-packages/jax/test_util.py", line 144, in check_vjp
    check_close(ip, ip_expected, atol=atol, rtol=rtol)
  File "/Users/hessian/venv/lib/python3.6/site-packages/jax/test_util.py", line 89, in check_close
    assert tree_all(tree_multimap(close, xs, ys)), '\n{} != \n{}'.format(xs, ys)
AssertionError: 
55.438952303284566 != 
55.43906353281106

======================================================================
FAIL: test_gpu_torsions (__main__.TestPeriodicTorsion)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "timemachine/jax_functionals/loss_of_precision_repro.py", line 174, in test_gpu_torsions
    check_grads(ref_nrg.angles, (conf,), order=2) # fails
  File "/Users/hessian/venv/lib/python3.6/site-packages/jax/test_util.py", line 153, in check_grads
    check_grads(f_vjp, args, order - 1, atol=atol, rtol=rtol, eps=eps)
  File "/Users/hessian/venv/lib/python3.6/site-packages/jax/test_util.py", line 160, in check_grads
    check_vjp(f, partial(api.vjp, f), args, atol, rtol, eps)
  File "/Users/hessian/venv/lib/python3.6/site-packages/jax/test_util.py", line 144, in check_vjp
    check_close(ip, ip_expected, atol=atol, rtol=rtol)
  File "/Users/hessian/venv/lib/python3.6/site-packages/jax/test_util.py", line 89, in check_close
    assert tree_all(tree_multimap(close, xs, ys)), '\n{} != \n{}'.format(xs, ys)
AssertionError: 
0.381797537133656 != 
0.381794076418857
```

**Comment by proteneer** [comment]
To add: this behavior is not observed in autograd.

**Comment by mattjj** [comment]
This is likely an XLA issue and not a JAX issue, in that there may not be much we can do on the front-end side. It's also true that XLA was designed for machine learning rather than high-precision applications.

But let's figure out what we might be able to do here!

@jlebar any gut reactions, or suggestions on how we can dig into this?

**Comment by jlebar** [comment]
This is likely a dupe of the existing bug for improving precision of reductions on CPU?

**Comment by proteneer** [comment]
The other issue (#585) is a sum of a million elements:
``` python
np.sum(np.ones(1000000) * 0.1).copy()
```

I'm no where near anything remotely as large number of items.

**Comment by proteneer** [comment]
To add: the results don't seem to change much for float32 - the original test case was for float64 - should I'd assumed would be free of errors of this magnitude.

**Comment by mattjj** [comment]
True but also you're not seeing as big of a drop in precision. The size of the precision drop is relative to the size of the inputs.

Is this a dupe if it's on the GPU backend? @proteneer is this on CPU or GPU?

**Comment by proteneer** [comment]
This on the CPU

**Comment by proteneer** [comment]
Note that this is also complicated by the fact that the errors are present only on the 2nd order derivatives, and not the 0th or 1st order derivatives.

**Comment by mattjj** [comment]
I suspect it is a duplicate then, though I’m not sure how to test that hypothesis (other than to wait for a resolution to #585 and see if this issue persists too).

**Comment by proteneer** [comment]
Wait a minute - something weird is going on. I didn't do my 32-bit test correctly previously.

If I turn off the first line:

``` python
from jax.config import config; config.update("jax_enable_x64", True)
```

Then my tests actually pass. There's something weird about turning _on_ 64bit that makes it worse.

**Comment by proteneer** [comment]
This is a finite difference step size discretization error. Changing the eps to 1e-5 and 1-e7 fixes the issue. It looks like the default was set to 1e-4, which is far too big of a step size for these types of applications.

``` python
from jax.config import config; config.update("jax_enable_x64", True)

import jax
import jax.numpy as np
import numpy as anp
from jax.test_util import check_grads
import unittest

class PeriodicTorsion():

    def __init__(self,
        torsion_idxs,
        param_idxs):
        """
        This implements a periodic torsional potential expanded out into three terms:

        V(a) = k0*(1+cos(1 * a - t0)) + k1*(1+cos(2 * a - t1)) + k2*(1+cos(3 * a - t2))

        Parameters:
        -----------
        torsion_idxs: [num_torsions, 4] np.array
            each element (a, b, c, d) is a torsion of four atoms, defined as
            as the angle of the plane defined by the three bond vectors a-b, b-c, c-d. 

        param_idxs: [num_torsions, 6] np.array
            each element (k, phase, periodicity) maps into params for angle constants and ideal angles

        """
        self.torsion_idxs = torsion_idxs
        self.param_idxs = param_idxs

    @staticmethod
    def get_signed_angle(ci, cj, ck, cl):
        """
        The torsion angle between two planes should be periodic but not
        necessarily symmetric. We use an identical but numerically stable arctan2
        implementation as opposed to the OpenMM energy function to avoid a
        singularity when the angle is zero. This is different from the CHARMM convention
        of symmetric torsions.
        """

        # Taken from the wikipedia arctan2 implementation:
        # https://en.wikipedia.org/wiki/Dihedral_angle

        rij = ci - cj
        rkj = ck - cj
        rkl = ck - cl

        n1 = np.cross(rij, rkj)
        n2 = np.cross(rkj, rkl)


        lhs = np.linalg.norm(n1, axis=-1)
        rhs = np.linalg.norm(n2, axis=-1)
        bot = lhs * rhs

        y = np.sum(np.multiply(np.cross(n1, n2), rkj/np.linalg.norm(rkj, axis=-1, keepdims=True)), axis=-1)
        x = np.sum(np.multiply(n1, n2), -1)

        return np.arctan2(y, x)

    def angles(self, conf):
        ci = conf[self.torsion_idxs[:, 0]]
        cj = conf[self.torsion_idxs[:, 1]]
        ck = conf[self.torsion_idxs[:, 2]]
        cl = conf[self.torsion_idxs[:, 3]]

        angle = self.get_signed_angle(ci, cj, ck, cl)
        return angle

    def energy(self, conf, params):
        """
        Compute the torsional energy.
        """
        ci = conf[self.torsion_idxs[:, 0]]
        cj = conf[self.torsion_idxs[:, 1]]
        ck = conf[self.torsion_idxs[:, 2]]
        cl = conf[self.torsion_idxs[:, 3]]

        ks = params[self.param_idxs[:, 0]]
        phase = params[self.param_idxs[:, 1]]
        period = params[self.param_idxs[:, 2]]
        angle = self.get_signed_angle(ci, cj, ck, cl)
        nrg = ks*(1+np.cos(period * angle - phase))
        return np.sum(nrg, axis=-1)


class TestPeriodicTorsion(unittest.TestCase):

    def setUp(self):
        self.conformers = anp.array([
            [[-0.6000563454193615, 0.376172954382274 ,-0.2487295756125901],
             [ 0.561317027011325 , 0.2066950040043141, 0.3670430960815993],
             [-1.187055522272264 ,-0.3415864358441354, 0.0871382207830652],
             [ 0.9399773448903637,-0.6888774474110431, 0.2104211949995816]],
            [[-0.6000563454193615, 0.376172954382274 ,-0.2487295756125901],
             [ 0.5613170270113252, 0.2066950040043142, 0.3670430960815993],
             [-1.187055522272264 ,-0.3415864358441354, 0.0871382207830652],
             [ 1.283345455745044 ,-0.0356257425880843,-0.2573923896494185]],
            [[-0.6000563454193615, 0.376172954382274 ,-0.2487295756125901],
             [ 0.561317027011325 , 0.2066950040043142, 0.3670430960815992],
             [-1.187055522272264 ,-0.3415864358441354, 0.0871382207830652],
             [ 1.263820400176392 , 0.7964992122869241, 0.0084568741589791]],
            [[-0.6000563454193615, 0.376172954382274 ,-0.2487295756125901],
             [ 0.5613170270113252, 0.2066950040043142, 0.3670430960815992],
             [-1.187055522272264 ,-0.3415864358441354, 0.0871382207830652],
             [ 0.8993534242298198, 1.042445571242743 , 0.7635483993060286]],
            [[-0.6000563454193615, 0.376172954382274 ,-0.2487295756125901],
             [ 0.5613170270113255, 0.2066950040043142, 0.3670430960815993],
             [-1.187055522272264 ,-0.3415864358441354, 0.0871382207830652],
             [ 0.5250337847650304, 0.476091386095139 , 1.3136545198545133]],
            [[-0.6000563454193615, 0.376172954382274 ,-0.2487295756125901],
             [ 0.5613170270113255, 0.2066950040043141, 0.3670430960815993],
             [-1.187055522272264 ,-0.3415864358441354, 0.0871382207830652],
             [ 0.485009232042489 ,-0.3818599172073237, 1.1530102055165103]],
            ], dtype=anp.float64)

    def test_gpu_torsions(self):
        """
        Test agreement of torsions with OpenMM's implementation of torsion terms.
        """
        torsion_idxs = anp.array([
            [0, 1, 2, 3],
            # [0, 1, 2, 3],
            # [0, 1, 2, 3],
        ], dtype=anp.int32)

        params_np = anp.array([
            2.3, # k0
            5.4, # k1
            9.0, # k2
            0.0, # t0
            3.0, # t1
            5.8, # t2
            1.0, # n0
            2.0, # n1
            3.0  # n2
        ], dtype=anp.float64)

        param_idxs = anp.array([
            # [0, 3, 6], # works
            [1, 4, 7], # fails
            # [2, 5, 8] # fails
        ], dtype=anp.int32)

        ref_nrg = PeriodicTorsion(
            param_idxs=param_idxs,
            torsion_idxs=torsion_idxs
        )

        for conf_idx, conf in enumerate(self.conformers):
            nrg = ref_nrg.energy(conf, params_np)
            angles = ref_nrg.angles(conf)

            print(nrg, angles)

            # first order derivatives pass
            check_grads(ref_nrg.angles, (conf,), order=1, eps=1e-5) # passes
            check_grads(ref_nrg.energy, (conf, params_np), order=1, eps=1e-5) # passes

            # second order derivatives fail miserably, both of these should pass in 64bit
            check_grads(ref_nrg.angles, (conf,), order=2, eps=1e-7) # fails
            check_grads(ref_nrg.energy, (conf, params_np), order=2, eps=1e-7) # fails




if __name__ == "__main__":
    unittest.main()
```

**Comment by proteneer** [comment]
@mattjj If you also agree that this is a finite difference step-size problem then I'd be happy to close this issue.

**Comment by mattjj** [comment]
Interesting! That sounds plausible. Finite differences is tricky because a smaller epsilon means smaller approximation error but bigger numerical error (because it involves dividing by a small number). If the test passes for any epsilon then it's likely okay.

There are numerical methods to reduce the approximation error, like Richardson extrapolation. That would be a nice thing to add to the test suite.

I think it makes sense to close the issue.

**Comment by proteneer** [comment]
Thanks - I think we're in agreement here.

**Comment by proteneer** [comment]
@mattjj  For pedagogy's sake, here's an example that fails for ANY finite difference step size all the way from 1e-1 to 1e-16:

``` python
from jax.config import config; config.update("jax_enable_x64", True)

import jax
import jax.numpy as np
import numpy as anp
from jax.test_util import check_grads
import unittest

class PeriodicTorsion():

    def __init__(self,
        torsion_idxs,
        param_idxs):
        """
        This implements a periodic torsional potential expanded out into three terms:

        V(a) = k0*(1+cos(1 * a - t0)) + k1*(1+cos(2 * a - t1)) + k2*(1+cos(3 * a - t2))

        Parameters:
        -----------
        torsion_idxs: [num_torsions, 4] np.array
            each element (a, b, c, d) is a torsion of four atoms, defined as
            as the angle of the plane defined by the three bond vectors a-b, b-c, c-d. 

        param_idxs: [num_torsions, 6] np.array
            each element (k, phase, periodicity) maps into params for angle constants and ideal angles

        """
        self.torsion_idxs = torsion_idxs
        self.param_idxs = param_idxs

    @staticmethod
    def get_signed_angle(ci, cj, ck, cl):
        """
        The torsion angle between two planes should be periodic but not
        necessarily symmetric. We use an identical but numerically stable arctan2
        implementation as opposed to the OpenMM energy function to avoid a
        singularity when the angle is zero. This is different from the CHARMM convention
        of symmetric torsions.
        """

        # Taken from the wikipedia arctan2 implementation:
        # https://en.wikipedia.org/wiki/Dihedral_angle

        rij = ci - cj
        rkj = ck - cj
        rkl = ck - cl

        n1 = np.cross(rij, rkj)
        n2 = np.cross(rkj, rkl)


        lhs = np.linalg.norm(n1, axis=-1)
        rhs = np.linalg.norm(n2, axis=-1)
        bot = lhs * rhs

        y = np.sum(np.multiply(np.cross(n1, n2), rkj/np.linalg.norm(rkj, axis=-1, keepdims=True)), axis=-1)
        x = np.sum(np.multiply(n1, n2), -1)

        return np.arctan2(y, x)

    def angles(self, conf):
        ci = conf[self.torsion_idxs[:, 0]]
        cj = conf[self.torsion_idxs[:, 1]]
        ck = conf[self.torsion_idxs[:, 2]]
        cl = conf[self.torsion_idxs[:, 3]]

        angle = self.get_signed_angle(ci, cj, ck, cl)
        return angle

    def energy(self, conf, params):
        """
        Compute the torsional energy.
        """
        ci = conf[self.torsion_idxs[:, 0]]
        cj = conf[self.torsion_idxs[:, 1]]
        ck = conf[self.torsion_idxs[:, 2]]
        cl = conf[self.torsion_idxs[:, 3]]

        ks = params[self.param_idxs[:, 0]]
        phase = params[self.param_idxs[:, 1]]
        period = params[self.param_idxs[:, 2]]
        angle = self.get_signed_angle(ci, cj, ck, cl)
        nrg = ks*(1+np.cos(period * angle - phase))
        return np.sum(nrg, axis=-1)


class TestPeriodicTorsion(unittest.TestCase):

    def setUp(self):
        self.nan_conformers = np.array([
            [[-0.6000563454193615, 0.376172954382274 ,-0.2487295756125901],
             [ 0.561317027011325 , 0.206695004004314 , 0.3670430960815994],
             [-1.187055522272264 ,-0.3415864358441354, 0.0871382207830652],
             [ 0.5494071252089705,-0.5626592973923106, 0.9817919758125693]],
            ], dtype=np.float64)

    def test_gpu_torsions(self):
        """
        Test agreement of torsions with OpenMM's implementation of torsion terms.
        """
        torsion_idxs = anp.array([
            [0, 1, 2, 3],
            [0, 1, 2, 3],
            [0, 1, 2, 3],
        ], dtype=anp.int32)

        params_np = anp.array([
            2.3, # k0
            5.4, # k1
            9.0, # k2
            0.0, # t0
            3.0, # t1
            5.8, # t2
            1.0, # n0
            2.0, # n1
            3.0  # n2
        ], dtype=anp.float64)

        param_idxs = anp.array([
            [0, 3, 6], # works
            [1, 4, 7], # fails
            [2, 5, 8] # fails
        ], dtype=anp.int32)

        ref_nrg = PeriodicTorsion(
            param_idxs=param_idxs,
            torsion_idxs=torsion_idxs
        )

        for conf_idx, conf in enumerate(self.nan_conformers):
            nrg = ref_nrg.energy(conf, params_np)
            angles = ref_nrg.angles(conf)

            print(nrg, angles)

            # first order derivatives pass
            check_grads(ref_nrg.angles, (conf,), order=1, eps=1e-16) # fails for ANY choice of eps from 1e-1 to 1e-16



if __name__ == "__main__":
    unittest.main()
```

**Comment by mattjj** [comment]
Does Autograd+NumPy have better behavior? That might not be trivial to check, since the `check_grads` function is slightly different there...

Do you consider this a bug?

**Comment by proteneer** [comment]
I do not see this behavior with autograd. This seems to be jax-specific.

**Comment by proteneer** [comment]
```
import autograd
import autograd.numpy as np
from autograd.test_util import check_grads

import numpy as anp
import unittest

class PeriodicTorsion():

    def __init__(self,
        torsion_idxs,
        param_idxs):
        """
        This implements a periodic torsional potential expanded out into three terms:

        V(a) = k0*(1+cos(1 * a - t0)) + k1*(1+cos(2 * a - t1)) + k2*(1+cos(3 * a - t2))

        Parameters:
        -----------
        torsion_idxs: [num_torsions, 4] np.array
            each element (a, b, c, d) is a torsion of four atoms, defined as
            as the angle of the plane defined by the three bond vectors a-b, b-c, c-d. 

        param_idxs: [num_torsions, 6] np.array
            each element (k, phase, periodicity) maps into params for angle constants and ideal angles

        """
        self.torsion_idxs = torsion_idxs
        self.param_idxs = param_idxs

    @staticmethod
    def get_signed_angle(ci, cj, ck, cl):
        """
        The torsion angle between two planes should be periodic but not
        necessarily symmetric. We use an identical but numerically stable arctan2
        implementation as opposed to the OpenMM energy function to avoid a
        singularity when the angle is zero. This is different from the CHARMM convention
        of symmetric torsions.
        """

        # Taken from the wikipedia arctan2 implementation:
        # https://en.wikipedia.org/wiki/Dihedral_angle

        rij = ci - cj
        rkj = ck - cj
        rkl = ck - cl

        n1 = np.cross(rij, rkj)
        n2 = np.cross(rkj, rkl)


        lhs = np.linalg.norm(n1, axis=-1)
        rhs = np.linalg.norm(n2, axis=-1)
        bot = lhs * rhs

        y = np.sum(np.multiply(np.cross(n1, n2), rkj/np.linalg.norm(rkj, axis=-1, keepdims=True)), axis=-1)
        x = np.sum(np.multiply(n1, n2), -1)

        return np.arctan2(y, x)

    def angles(self, conf):
        ci = conf[self.torsion_idxs[:, 0]]
        cj = conf[self.torsion_idxs[:, 1]]
        ck = conf[self.torsion_idxs[:, 2]]
        cl = conf[self.torsion_idxs[:, 3]]

        angle = self.get_signed_angle(ci, cj, ck, cl)
        return angle

    def energy(self, conf, params):
        """
        Compute the torsional energy.
        """
        ci = conf[self.torsion_idxs[:, 0]]
        cj = conf[self.torsion_idxs[:, 1]]
        ck = conf[self.torsion_idxs[:, 2]]
        cl = conf[self.torsion_idxs[:, 3]]

        ks = params[self.param_idxs[:, 0]]
        phase = params[self.param_idxs[:, 1]]
        period = params[self.param_idxs[:, 2]]
        angle = self.get_signed_angle(ci, cj, ck, cl)
        nrg = ks*(1+np.cos(period * angle - phase))
        return np.sum(nrg, axis=-1)


class TestPeriodicTorsion(unittest.TestCase):

    def setUp(self):
        self.nan_conformers = np.array([
            [[-0.6000563454193615, 0.376172954382274 ,-0.2487295756125901],
             [ 0.561317027011325 , 0.206695004004314 , 0.3670430960815994],
             [-1.187055522272264 ,-0.3415864358441354, 0.0871382207830652],
             [ 0.5494071252089705,-0.5626592973923106, 0.9817919758125693]],
            ], dtype=np.float64)

    def test_gpu_torsions(self):
        """
        Test agreement of torsions with OpenMM's implementation of torsion terms.
        """
        torsion_idxs = anp.array([
            [0, 1, 2, 3],
            [0, 1, 2, 3],
            [0, 1, 2, 3],
        ], dtype=anp.int32)

        params_np = anp.array([
            2.3, # k0
            5.4, # k1
            9.0, # k2
            0.0, # t0
            3.0, # t1
            5.8, # t2
            1.0, # n0
            2.0, # n1
            3.0  # n2
        ], dtype=anp.float64)

        param_idxs = anp.array([
            [0, 3, 6], # works
            [1, 4, 7], # fails
            [2, 5, 8] # fails
        ], dtype=anp.int32)

        ref_nrg = PeriodicTorsion(
            param_idxs=param_idxs,
            torsion_idxs=torsion_idxs
        )

        for conf_idx, conf in enumerate(self.nan_conformers):
            nrg = ref_nrg.energy(conf, params_np)
            angles = ref_nrg.angles(conf)

            print(nrg, angles)

            # first order derivatives pass
            check_grads(ref_nrg.angles, (conf,), order=1, eps=1e-8) # fails for ANY choice of eps from 1e-1 to 1e-16
            check_grads(ref_nrg.angles, (conf,), order=2, eps=1e-8) # fails for ANY choice of eps from 1e-1 to 1e-16



if __name__ == "__main__":
    unittest.main()
```

**Comment by gnool** [comment]
@mattjj Trying to get your opinion on comparing JAX jacobian results vs finite difference results here as it seems at least remotely related to this issue.

I've prepared a cost function in two versions - JAX (single precision) and numpy (double precision). The derivatives for the former is (obviously) auto differentiation, while the latter is finite difference. I noticed that at eps=1e-3, both results compare well. If I reduce eps further (say to 1e-5 or 1e-7), the results become really different. 

We know that in principle with a small step size, the estimated gradient should be better but in practice, numerical errors such as that associated by division by small number kick in. In such a case, how likely is it that all the difference observed between the two versions is entirely due to the error in finite difference method? Is double precision used in the numpy version not sufficient to perform the calculation accurately? I'm also curious as to which decimal of the JAX results that I can trust when we use single precision.

Is there a third independent way I can use to assess the accuracy of the two approaches above?

**Comment by proteneer** [comment]
Use the complex step derivative: http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.141.8002&rep=rep1&type=pdf

The implementation is trivial, just cast the input types to a complex number and follow the instructions in the paper above.

**Comment by proteneer** [comment]
Note that this assumes that your function is analytic (in the complex sense)

**Comment by gnool** [comment]
@proteneer Actually, I observe that if I turn on double precision for JAX, the results then become similar again with those from finite difference. So it does seem like in my case above the error was coming from JAX due to single precision? Have to say I am a bit reluctant to admit that I really need double precision in JAX... (was hoping I could save some computational resource by using single precision)

**Comment by diegoferigo** [comment]
I found this issue during a debugging session of my code due to failures occurring after calling `jax.test_util.check_grads` on my piece of computation. TL;DR, by chance I switched to 64 bits and magically the check started passing. For all those having problems with their gradients, as a first attempt I'd suggest to enable 64 bits support since a lower floating point precision might be the original cause.

> I've prepared a cost function in two versions - JAX (single precision) and numpy (double precision). The derivatives for the former is (obviously) auto differentiation, while the latter is finite difference. I noticed that at eps=1e-3, both results compare well. If I reduce eps further (say to 1e-5 or 1e-7), the results become really different.

Note that the precision of the active floating point arithmetic matters on the selection of $\epsilon$. Referring [Numerical_Differentiation#Step_size](https://en.wikipedia.org/wiki/Numerical_differentiation#Step_size), a possibility to calculate the step size is the following:

```python
eps = jnp.finfo(jnp.array(0.0, dtype=float)).resolution ** (1 / 3)
```

It automatically picks the active floating point configuration of jax and computes the right step size for `check_grads`.
